### DISCUSSION TITLE
如何支持数据库实体定义雪花ID并自动赋值

### DISCUSSION THREAD

**Discussion by fzwu** [luoyunchong/lin-cms-dotnetcore#31] [Q&A] [answered]
目前使用FullAuditEntity，long类型的需要设置为非自增且需要手动赋值，不支持自动获取雪花ID赋值，建议能够支持雪花ID赋值

**Comment by luoyunchong** [comment]
mysql下主键自增也能赋值，

- 包
```xml
<PackageReference Include="Yitter.IdGenerator" Version="1.0.14" />
```

-（新增）FreeSqlBuilder增加优先级，按顺序覆盖的逻辑，最后的那个优先级最高
```cs
UseMappingPriority(MappingPriorityType.FluentApi, MappingPriorityType.Attribute, MappingPriorityType.Aop)
```
- 配置fsql Aop
```cs
    var options = new IdGeneratorOptions(1);
        // options.WorkerIdBitLength = 10; // 默认值6，限定 WorkerId 最大值为2^6-1，即默认最多支持64个节点。
        // options.SeqBitLength = 6; // 默认值6，限制每毫秒生成的ID个数。若生成速度超过5万个/秒，建议加大 SeqBitLength 到 10。
        // options.BaseTime = Your_Base_Time; // 如果要兼容老系统的雪花算法，此处应设置为老系统的BaseTime。
        // ...... 其它参数参考 IdGeneratorOptions 定义。
        YitIdHelper.SetIdGenerator(options);
        // 保存参数（务必调用，否则参数设置不生效）

        fsql.Aop.AuditValue += (s, e) =>
        {
            if (e.Column.CsType == typeof(long) && e.Property.Name == "Id" && e.Value?.ToString() == "0")
            {
                e.Value = YitIdHelper.NextId();
            }
        };
        fsql.Aop.ConfigEntityProperty += (s, e) =>
        {
            if ( e.Property.Name == "Id")
                e.ModifyResult.IsIdentity = false;
        };
```

![image](https://user-images.githubusercontent.com/18613266/223188983-a073ae7a-b76a-41f9-9164-5dc1c15fe47b.png)

**Reply by fzwu** [reply]
好的 ，多谢

**Reply by fzwu** [reply]
fsql.Aop.AuditValue += (s, e) =>
                {
                    if (e.Column.CsType == typeof(long) && e.Property.Name == "id" && e.Value?.ToString() == "0")
                    {
                        e.Value = YitIdHelper.NextId();
                    }
                };

                fsql.Aop.ConfigEntityProperty += (s, e) =>
                {
                    if (e.Property.Name == "id")
                        e.ModifyResult.IsIdentity = false;
                };

                fsql.GlobalFilter.Apply<ISoftDelete>("IsDeleted", a => a.IsDeleted == false);

                fsql.UseJsonMap();
                
                return fsql;

我试了一下，但是这里报错了  

2023-03-07 14:27:52.913 +08:00 [ERR] Message:不可添加，自增属性有值：(1632612029224427520, , , 0001/1/1 0:00:00, , , , , , , False, Admin, 系统管理员, True, 0),StackTrace:   at FreeSql.DbSet`1.CanAdd(TEntity data, Boolean isThrow)
   at FreeSql.DbSet`1.CanAdd(IEnumerable`1 data, Boolean isThrow)
   at FreeSql.DbSet`1.AddRange(IEnumerable`1 data)
   at FreeSql.BaseRepository`1.Insert(IEnumerable`1 entitys)
   at FreeSql.Extensions.EfCoreFluentApi.EfCoreTableFluent`1.<>c__DisplayClass17_0.<HasData>b__1(Object s, SyncStructureAfterEventArgs e)
   at FreeSql.Internal.CommonProvider.CodeFirstProvider.SyncStructure(TypeAndName[] objects)
   at FreeSql.Internal.CommonProvider.CodeFirstProvider.SyncStructure(Type[] entityTypes)

**Reply by luoyunchong** [reply]
if ( e.Property.Name == "Id")

**Reply by fzwu** [reply]
也是一样的错误，通过下面的方式是可以的
        [Column(IsIdentity = false)]
        public override long Id { get; set; }

**Reply by luoyunchong** [reply]
还以为`   e.Column.Attribute.IsIdentity = false;`没效果，本来写了又删了，加一行
```cs
 fsql.Aop.AuditValue += (s, e) =>
    {
        if (e.Column.CsType == typeof(long) && e.Property.Name == "Id" && e.Value?.ToString() == "0")
        {
            e.Value = YitIdHelper.NextId();
            e.Column.Attribute.IsIdentity = false;
        }
    };
```

**Comment by luoyunchong** [comment, marked as answer]
或第二种方式

FreeSqlBuilder配置优先级，按顺序覆盖的逻辑，最后的那个优先级最高
```cs
UseMappingPriority(MappingPriorityType.FluentApi, MappingPriorityType.Attribute, MappingPriorityType.Aop)
```
```cs
  fsql.Aop.AuditValue += (s, e) =>
  {
      if (e.Column.CsType == typeof(long) && e.Property.Name == "Id" && e.Value?.ToString() == "0")
      {
          e.Value = YitIdHelper.NextId();
      }
  };
  fsql.Aop.ConfigEntityProperty += (s, e) =>
  {
      if ( e.Property.Name == "Id")
          e.ModifyResult.IsIdentity = false;
  };
```

**Reply by fzwu** [reply]
好的，这两种方式是可以的，多谢
